<a href="https://colab.research.google.com/github/ridwan-bigdata/Pagila-SQL-task/blob/main/cereal_calorie_analysis_numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Numpy Library Exercises

In [1]:
import numpy as np

**1.** Given a random array, change the sign of the elements whose values ​​are between 3 and 8

In [4]:
#Mathematically, we are trying to write the values within the range 3<= x<= 8.
#That is if the element x is between 3 and 8 inclusive, multiply by -1. Otherwise keep the value unchanged.

arr= np.array([1, 4, 7, 10, 3, 8, 6, 2])
arr[(arr > 3) & (arr < 8)] *= -1
print(arr)

[ 1 -4 -7 10  3  8 -6  2]


**2.** Replace the maximum element of a random array with 0

In [5]:
arr = np.random.randint(1, 20, 10)

arr[arr == arr.max()] = 0

print(arr)

[ 5  6 17 11 17  0  8 11  0 10]


**3.** Construct a direct product of arrays (all combinations with each element). The input is a two-dimensional array.

In [6]:
from itertools import product

arr = [
    [1, 2],
    [3, 4],
    [5, 6]
]

result = list(product(*arr))

print(result)

[(1, 3, 5), (1, 3, 6), (1, 4, 5), (1, 4, 6), (2, 3, 5), (2, 3, 6), (2, 4, 5), (2, 4, 6)]


**4.** Given two arrays A (8x3) and B (2x2), find the rows in A that contain elements from each row in B, regardless of the order of the elements in B.

In [9]:
# Example arrays
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [1, 4, 2],
    [7, 8, 9],
    [2, 1, 5],
    [6, 4, 5],
    [3, 1, 2],
    [8, 7, 6]
])

B = np.array([
    [1, 2],
    [4, 5]
])

# Find rows in A containing elements of each row in B
result = []

for b in B:
    mask = np.all(np.isin(b, A), axis=0)
    rows = [i for i, row in enumerate(A) if np.all(np.isin(b, row))]
    result.append(rows)

print(result)

[[0, 2, 4, 6], [1, 5]]


**5.** Given a 10x3 matrix, find rows of unequal values ​​(for example, row [2,2,3] remains, row [3,3,3] is deleted)

In [13]:
# Example 10x3 matrix
A = np.array([
    [2,2,3],
    [3,3,3],
    [4,5,6],
    [7,7,7],
    [1,2,1],
    [9,9,8],
    [5,5,5],
    [3,4,3],
    [6,6,7],
    [8,8,8]
])

# Keep rows where elements are not all equal
result = A[~np.all(A == A[:, [0]], axis=1)]

print(result)

[[2 2 3]
 [4 5 6]
 [1 2 1]
 [9 9 8]
 [3 4 3]
 [6 6 7]]


**6.** Given a two-dimensional array, remove those lines that are repeated

In [14]:
# Example 2D array
A = np.array([
    [1,2,3],
    [4,5,6],
    [1,2,3],
    [7,8,9],
    [4,5,6],
    [10,11,12]
])

# Remove repeated rows
result = np.unique(A, axis=0)

print(result)

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


______
______

For each of the following problems (1-5), you need to provide two implementations – one without using NumPy (assume that where the input or output should be NumPy arrays, there will simply be lists), and the other fully vectorized using NumPy (without using Python loops/map/list comprehensions). Note 1.__ You can assume that all the specified objects are non-empty (for example, in __problem 1__, there are non-zero elements on the diagonal of the matrix). Note 2.__ For most problems, the solution takes no more than 1-2 lines.

___

* __Problem 1__: Calculate the product of nonzero elements on the diagonal of a rectangular matrix. For example, for X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) the answer is 3.

In [20]:
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
# Without Numpy
import math
result = math.prod(X[i][i] for i in range(min(len(X), len(X[0]))) if X[i][i] != 0)
#print(result)

# With Numpy
result = np.prod(np.diag(X)[np.diag(X) != 0])
print(result)

3


* __Problem 2__: Given two vectors x and y, check whether they define the same multiset. For example, for x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]), the answer is True.

In [25]:
# Without Numpy

x = [1,2,2,4]
y = [4,2,1,2]

from collections import Counter
result = Counter(x) == Counter(y)
#print(result)

#With Numpy

result = np.array_equal(np.sort(x), np.sort(y))
print(result)

True


* __Problem 3__: Find the maximum element in the vector x among elements preceded by zero. For example, for x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) the answer is 5.

In [30]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
# Without Numpy
result = max(x[i] for i in range(1, len(x)) if x[i-1] == 0 and x[i] != 0)
#print(result)

# With Numpy
result = np.max(x[1:][x[:-1] == 0])
print(result)

5


* __Problem 4__: Implement run-length encoding. For a given vector x, return a tuple of two vectors of the same length. The first contains numbers, and the second contains the number of times they should be repeated. For example, for x = np.array([2, 2, 2, 3, 3, 3, 5]), the answer is (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [35]:
x = np.array([2, 2, 2, 3, 3, 3, 5])

# Without Numpy

values, counts = [], []
for v in x:
    if not values or v != values[-1]:
        values.append(v); counts.append(1)
    else:
        counts[-1] += 1
# print(values, counts)

#With Numpy

idx = np.flatnonzero(np.r_[True, x[1:] != x[:-1]])
values = x[idx]
counts = np.diff(np.r_[idx, len(x)])
print(values, counts)

[2 3 5] [3 3 1]


* __Problem 5__: Given two samples of objects, X and Y, calculate the matrix of Euclidean distances between the objects. Compare this with the scipy.spatial.distance.cdist function in terms of speed.

In [47]:
# Without Numpy

import math

X = [[1, 2], [3, 4]]
Y = [[5, 6], [7, 8]]

D = [[math.sqrt(sum((xi - yi) ** 2 for xi, yi in zip(x, y))) for y in Y] for x in X]
# print(D)

# With Numpy
X_np = np.array(X)
Y_np = np.array(Y)
D_np = np.sqrt((X_np**2).sum(axis=1)[:, None] + (Y_np**2).sum(axis=1) - 2 * X_np @ Y_np.T)
#print(D_np)

#SciPY implementation
from scipy.spatial.distance import cdist

D = cdist(X, Y, metric='euclidean')
print(D)

[[5.65685425 8.48528137]
 [2.82842712 5.65685425]]


_______
________

* #### __Task 6__: CrunchieMunchies __*__ You work in the marketing department of MyCrunch, a food company developing a new type of delicious, healthy cereal called **CrunchieMunchies**. You want to demonstrate to consumers how healthy your cereal is compared to other leading brands, so you've collected nutrition data from several different competitors. Your task is to use Numpy calculations to analyze this data and prove that your **CrunchieMunchies** are the healthiest choice for consumers.

In [49]:
import numpy as np

1. View the cereal.csv file. This file contains calorie counts for various brands of cereal. Download the data from the file and save it as calorie_stats.

In [54]:
import os
print(os.getcwd())

/content


In [55]:
from google.colab import files
uploaded = files.upload()

Saving cereal.csv to cereal.csv


In [59]:
import pandas as pd

calorie_stats = pd.read_csv("cereal.csv")
calorie_stats.head(5)

,70,120,70,50,110,110,110.1,130,90,90.1,...,110.23,90.6,110.24,140.2,100.14,110.25,110.26,100.15,100.16,110.27


2. One serving of CrunchieMunchies contains 60 calories. How much higher is the average calorie count of your competitors? Save the answer in the variable average_calories and print it in the terminal.

In [63]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")
average_calories = np.mean(calorie_stats) - 60

print(average_calories)

46.883116883116884


3. Does the average calorie count accurately reflect the distribution of the data set? Let's sort the data and see. Sort the data and save the result in the calorie_stats_sorted variable. Print the sorted information.

In [65]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")
calorie_stats_sorted = np.sort(calorie_stats)

print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. It looks like most values ​​are above the mean. Let's see if the median is the most accurate measure of the data set. Calculate the median of the data set and save your answer in median_calories. Print the median so you can see how it compares to the mean.

In [69]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")
mean_calories = np.mean(calorie_stats)
print (mean_calories)

# mean_calories is 106.88

median_calories = np.median(calorie_stats)

print(median_calories)


106.88311688311688
110.0


5. While the median shows that at least half of our values ​​are over 100 calories, it would be more impressive to show that a significant portion of the competitors have a higher calorie count than CrunchieMunchies. Calculate the various percentiles and print them until you find the lowest percentile exceeding 60 calories. Store this value in the nth_percentile variable.

In [70]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")

for i in range(100):
    percentile = np.percentile(calorie_stats, i)
    if percentile > 60:
        nth_percentile = i
        break

print(nth_percentile)

4


6. While the percentile shows us that most competitors have much higher calorie counts, it's an inconvenient concept to use in marketing materials. Instead, let's calculate the percentage of cereals that contain more than 60 calories per serving. Save your answer in the more_calories variable and print it.

In [71]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")

more_calories = np.mean(calorie_stats > 60) * 100

print(more_calories)

96.1038961038961


7. That's a really high percentage. This will be very useful when we promote CrunchieMunchies. But one question is, how much variation is there in the data set? Can we generalize that most cereals contain around 100 calories, or is the range even greater? Calculate the magnitude of the variation by finding the standard deviation. Save your answer in calorie_std and print it to the terminal. How can we incorporate this value into our analysis?

In [72]:
calorie_stats = np.genfromtxt("cereal.csv", delimiter=",")

calorie_std = np.std(calorie_stats)

print(calorie_std)

19.35718533390827


8. Write a short paragraph summarizing your findings and how you believe this data could be used to Mycrunch's advantage in marketing CrunchieMunchies.

Answer: The analysis of the cereal calorie dataset shows that most competitor cereals contain significantly more calories than CrunchieMunchies. The average calorie count of competing cereals is well above 60 calories, and approximately 96% of cereals exceed 60 calories per serving, highlighting how much lighter CrunchieMunchies is compared to most options on the market. While the mean calorie count is around 107 calories, the median is also close to 100 calories, confirming that most cereals cluster around this higher range. Additionally, the standard deviation indicates that even with some variation, the majority of cereals still fall far above the 60-calorie level. This data can be used in marketing to position CrunchieMunchies as a low-calorie, healthier alternative, emphasizing that consumers can enjoy cereal with significantly fewer calories than the vast majority of competing brands.